ocaml/xenopsd/lib/xenops_server.ml

```diff
@@ -3367,12 +3448,41 @@ module VM = struct
                 Hashtbl.remove vgpu_receiver_sync vm_id
             )
         | None ->
-            let headers = Cohttp.Header.of_list [("User-agent", "xenopsd")] in
-            let response =
-              Cohttp.Response.make ~version:`HTTP_1_1 ~status:`Not_found
-                ~headers ()
-            in
-            Response.write (fun _ -> ()) response s
+            Response.write (fun _ -> ()) not_found_response s
+      )
+      ()
+
+  (* This handler /service/xenops/migrate-mem/id  receives a connection for
+     VM memory. *)
+  let receive_mem uri cookies socket context : unit =
+    let module Request = Cohttp.Request.Make (Cohttp_posix_io.Unbuffered_IO) in
+    let module Response = Cohttp.Response.Make (Cohttp_posix_io.Unbuffered_IO) in
+    let dbg = List.assoc "dbg" cookies in
+    Debug.with_thread_associated dbg
+      (fun () ->
+        let vm = basename (Uri.path uri) in
+        match context.transferred_fd with
+        | Some fd ->
+            debug "VM.receive_mem: passed fd %d" (Obj.magic fd) ;
+            debug "VM.receive_mem: vm=%s" vm ;
+            (* Store away the fd for VM_receive_memory/restore to use *)
+            let info = {mem_fd= fd; mem_channel= Event.new_channel ()} in
+            with_lock mem_receiver_sync_m (fun () ->
+                Hashtbl.add mem_receiver_sync vm info
+            ) ;
+            (* Inform the sender that everything is in place to start
+               save/restore *)
+            Xenops_migrate.(Handshake.send ~verbose:true fd Handshake.Success) ;
+            debug "VM.receive_mem: Synchronisation point 1-mem" ;
+            (* Keep the thread/connection open until the restore is done on the
+               other thread *)
+            Event.receive info.mem_channel |> Event.sync ;
```

Comment:
What if the migration fails, are we going to wait here forever, potentially "leaking" the file descriptor? I wonder whether this is causing the xenopsd file descriptor leak bug that we have elsewhere (not your new code, but the existing migration code in xenopsd)

Reply:
The signal from the main thread is sent in a finally handler that protects most of the code before it.